File: 106_Generate Parentheses.py
"""
Generate Parentheses
Return all combinations of well-formed parentheses for n pairs.
"""
from typing import List

def generate_parentheses(n: int) -> List[str]:
    res = []
    def dfs(cur, open_count, close_count):
        if len(cur) == 2*n:
            res.append(''.join(cur)); return
        if open_count < n:
            cur.append('('); dfs(cur, open_count+1, close_count); cur.pop()
        if close_count < open_count:
            cur.append(')'); dfs(cur, open_count, close_count+1); cur.pop()
    dfs([], 0, 0)
    return res

if __name__ == '__main__':
    print('Generate Parentheses n=3:', generate_parentheses(3))

# Time Complexity: Catalan number output size
# Space Complexity: O(n)

File: 106_Generate Parentheses.cpp
// Generate Parentheses
// Generate all combinations of well-formed parentheses for n pairs.
// Solution: backtracking tracking counts of open and close parens.
#include <bits/stdc++.h>
using namespace std;

vector<string> generateParenthesis(int n) {
    vector<string> res;
    function<void(string,int,int)> dfs = [&](string cur, int open, int close){
        if(cur.size() == 2*n){ res.push_back(cur); return; }
        if(open < n) dfs(cur + '(', open+1, close);
        if(close < open) dfs(cur + ')', open, close+1);
    };
    dfs("", 0, 0);
    return res;
}

int main(){
    auto r = generateParenthesis(3);
    cout<<"Generate Parentheses (n=3):\n";
    for(auto &s: r) cout<<s<<"\n";
    return 0;
}

// Time Complexity: Catalan number ~ O(4^n / (n*sqrt(n))) outputs
// Space Complexity: O(n) recursion + output
